src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs

```diff
@@ -43,253 +54,205 @@ protected AbstractUseAutoPropertyAnalyzer()
-            var option = context.GetAnalyzerOptions().PreferAutoProperties;
-            if (!option.Value)
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSymbolStartAction(context =>
             {
-                return;
-            }
-
-            var analysisResults = new List<AnalysisResult>();
-            var ineligibleFields = new HashSet<IFieldSymbol>();
+                var namedType = (INamedTypeSymbol)context.Symbol;
+                if (namedType.TypeKind is not TypeKind.Class and not TypeKind.Struct and not TypeKind.Module)
+                    return;
 
-            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
-            AnalyzeCompilationUnit(context, root, analysisResults);
+                var analysisResults = new ConcurrentQueue<AnalysisResult>();
+                context.RegisterSyntaxNodeAction(context => AnalyzePropertyDeclaration(context, namedType, analysisResults), PropertyDeclarationKind);
 
-            RegisterIneligibleFieldsAction(
-                analysisResults, ineligibleFields,
-                semanticModel.Compilation, cancellationToken);
-            Process(analysisResults, ineligibleFields, context);
-        }
+                var ineligibleFields = new ConcurrentSet<IFieldSymbol>();
+                var nonConstructorFieldWrites = new ConcurrentDictionary<IFieldSymbol, ConcurrentSet<SyntaxNode>>();
 
-        protected void AnalyzeProperty(
-            SemanticModelAnalysisContext context, TPropertyDeclaration propertyDeclaration, List<AnalysisResult> analysisResults)
+                context.RegisterCodeBlockStartAction<TSyntaxKind>(context =>
+                {
+                    RegisterIneligibleFieldsAction(ineligibleFields, context.SemanticModel, context.CodeBlock, context.CancellationToken);
+                    RegisterNonConstructorFieldWrites(nonConstructorFieldWrites, context.SemanticModel, context.CodeBlock, context.CancellationToken);
+                });
+
+                context.RegisterSymbolEndAction(context =>
+                    Process(analysisResults, ineligibleFields, nonConstructorFieldWrites, context));
+            }, SymbolKind.NamedType);
+
+        private void AnalyzePropertyDeclaration(
+            SyntaxNodeAnalysisContext context,
+            INamedTypeSymbol containingType,
+            ConcurrentQueue<AnalysisResult> analysisResults)
         {
             var cancellationToken = context.CancellationToken;
             var semanticModel = context.SemanticModel;
+            var compilation = semanticModel.Compilation;
 
+            var propertyDeclaration = (TPropertyDeclaration)context.Node;
             if (semanticModel.GetDeclaredSymbol(propertyDeclaration, cancellationToken) is not IPropertySymbol property)
-            {
                 return;
-            }
+
+            if (!containingType.Equals(property.ContainingType))
+                return;
```

Comment:
not sure if this check is necessary.  but i'm paranoid about being called back about a property in a *nested* type.  I'm not sure the rules about RegisterSyntaxNodeAction within a RegisterSymbolStartAction

Reply:
This shouldn't be necessary, and can be converted into an assert. Definitely suggest adding a unit test with property decls in both outer and nested type and making sure this is indeed true.